resources/net.go

```diff
@@ -0,0 +1,726 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/purpleidea/mgmt/recwatch"
+	"github.com/purpleidea/mgmt/util"
+
+	systemd "github.com/coreos/go-systemd/dbus" // change namespace
+	systemdUtil "github.com/coreos/go-systemd/util"
+	"github.com/godbus/dbus"
+	errwrap "github.com/pkg/errors"
+)
+
+const (
+	// IfacePrefix is used to identify unit files for managed links.
+	IfacePrefix = "mgmt-"
+	// IfaceUp is the up interface state
+	IfaceUp = "up"
+	// IfaceDown is the down interface state
+	IfaceDown = "down"
+
+	// DBusInterface is DBus's own DBus interface.
+	DBusInterface = "org.freedesktop.DBus"
+	// DBusAddMatch is the DBus method to filter signals.
+	DBusAddMatch = DBusInterface + ".AddMatch"
+	// DBusPropertiesInterface is the DBus interface used to get events.
+	DBusPropertiesInterface = DBusInterface + ".Properties"
+	// DBusPropertiesGet is the DBus method to get a property.
+	DBusPropertiesGet = DBusPropertiesInterface + ".Get"
+	// DBusNetworkdInterface is the networkd dbus interface.
+	DBusNetworkdInterface = "org.freedesktop.network1"
+	// DBusNetworkdLinkInterface is the networkd dbus link interface.
+	DBusNetworkdLinkInterface = DBusNetworkdInterface + ".Link"
+	// DBusLinkPath is the DBus source path for network interface events.
+	DBusLinkPath = "/org/freedesktop/network1/link/"
+	// DBusEavesdropSprint is the format of dbus eavesdrop arguments.
+	DBusEavesdropSprint = "type='signal', interface='%s', path='%s', eavesdrop='true'"
+	// DBusFailMode is the dbus mode used for restarting the network service
+	DBusFailMode = "fail"
+
+	// NetworkdUnitFileDir is the location of networkd unit files.
+	NetworkdUnitFileDir = "/etc/systemd/network/"
+	// NetworkdUnitFileExt is the file extension for networkd unit files.
+	NetworkdUnitFileExt = ".network"
+	// NetworkdService is the name of the networkd service.
+	NetworkdService = "systemd-networkd.service"
+
+	// ActiveState is the systemd-networkd service's active state.
+	ActiveState = "ActiveState"
+	// ActiveStateActive is the active service state.
+	ActiveStateActive = "active"
+	// ActiveStateReloading is the reloading service state.
+	ActiveStateReloading = "reloading"
+
+	// OpState is the operational state of the device.
+	OpState = "OperationalState"
+	// OpStateRoutable us the routable operational state.
+	OpStateRoutable = "routable"
+	// OpStateNoCarrier is the no-carrier operational state.
+	OpStateNoCarrier = "no-carrier"
+	// OpStateOff is the off operational state.
+	OpStateOff = "off"
+)
+
+func init() {
+	RegisterResource("net", func() Res { return &NetRes{} })
+}
+
+// NetRes is an systemd-networkd resource.
+type NetRes struct {
+	BaseRes `yaml:",inline"`
+	State   string        `yaml:"state"`   // up or down
+	IPv4    string        `yaml:"ipv4"`    // ipv4 address in cidr format
```

Comment:
Network interfaces can have a list of IP addresses. Don't change this right away, but let's discuss. Shouldn't we accept a list instead? In particular, what do we do if we only wanted mgmt to manage one IP address, but leave the others alone? Or do we only let it manage the whole thing or nothing?